File: app/src/main/java/com/bresiu/codechallenge/repository/Repository.java
package com.bresiu.codechallenge.repository;

import android.util.Log;
import androidx.lifecycle.LiveData;
import com.bresiu.codechallenge.data.entity.EntitiesCombined;
import com.bresiu.codechallenge.model.PostWithUserAddress;
import com.bresiu.codechallenge.model.Response;
import com.bresiu.codechallenge.repository.mappers.AlbumModelToEntityMapper;
import com.bresiu.codechallenge.repository.mappers.Mapper;
import com.bresiu.codechallenge.repository.mappers.PhotoModelToEntityMapper;
import com.bresiu.codechallenge.repository.mappers.PostModelToEntityMapper;
import com.bresiu.codechallenge.repository.mappers.UserCombinedModelToEntitiesMapper;
import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;
import java.util.List;
import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton public class Repository {
	private final DBRepository dbRepository;
	private final ApiRepository apiRepository;

	@Inject Repository(DBRepository dbRepository, ApiRepository apiRepository) {
		this.dbRepository = dbRepository;
		this.apiRepository = apiRepository;
	}

	public Observable<EntitiesCombined> fetchData() {
		return Observable.zip(apiRepository.getPosts(), apiRepository.getUsers(),
				apiRepository.getAlbums(), apiRepository.getPhotos(), Response::new)
				.observeOn(Schedulers.io())
				.map(response -> new EntitiesCombined(
						Mapper.INSTANCE.mapObject(response.users, new UserCombinedModelToEntitiesMapper()),
						Mapper.INSTANCE.mapList(response.posts, new PostModelToEntityMapper()),
						Mapper.INSTANCE.mapList(response.albums, new AlbumModelToEntityMapper()),
						Mapper.INSTANCE.mapList(response.photos, new PhotoModelToEntityMapper())))
				.subscribeOn(Schedulers.io());
	}

	public void saveData(EntitiesCombined entitiesCombined) {
		Log.d("BRS", "saveData");
		dbRepository.saveUsers(entitiesCombined.userCombined);
		dbRepository.savePosts(entitiesCombined.posts);
		dbRepository.saveAlbums(entitiesCombined.albums);
		dbRepository.savePhotos(entitiesCombined.photos);
	}

	public LiveData<List<PostWithUserAddress>> getLiveData() {
		return dbRepository.fetchPostsWithUserAddress();
	}

	public void deletePostById(long postId) {
		dbRepository.deletePostById(postId);
	}
}

File: app/src/main/java/com/bresiu/codechallenge/repository/Repository.kt
package com.bresiu.codechallenge.repository

import androidx.lifecycle.LiveData
import androidx.lifecycle.LiveDataReactiveStreams
import com.bresiu.codechallenge.data.entity.EntitiesCombined
import com.bresiu.codechallenge.model.*
import com.bresiu.codechallenge.repository.mappers.*
import io.reactivex.Observable
import io.reactivex.functions.Function4
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton


@Singleton
class Repository @Inject internal constructor(private val dbRepository: DBRepository,
                                              private val apiRepository: ApiRepository) {

  fun fetchData(): Observable<EntitiesCombined> {
    return Observable.zip(apiRepository.posts, apiRepository.users,
        apiRepository.albums, apiRepository.photos,
        Function4<List<Post>, List<User>, List<Album>, List<Photo>,
            Response> { posts, users, albums, photos ->
          Response(posts, users, albums, photos)
        }).observeOn(Schedulers.io())
        .map { response ->
          EntitiesCombined(
              Mapper.mapObject(response.users, UserCombinedModelToEntitiesMapper()),
              Mapper.mapList(response.posts, PostModelToEntityMapper()),
              Mapper.mapList(response.albums, AlbumModelToEntityMapper()),
              Mapper.mapList(response.photos, PhotoModelToEntityMapper()))
        }
        .subscribeOn(Schedulers.io())
  }

  fun deletePostById(postId: Long) {
    dbRepository.deletePostById(postId)
  }

  fun saveData(entitiesCombined: EntitiesCombined) {
    dbRepository.saveUsers(entitiesCombined.userCombined)
    dbRepository.savePosts(entitiesCombined.posts)
    dbRepository.saveAlbums(entitiesCombined.albums)
    dbRepository.savePhotos(entitiesCombined.photos)
  }

  fun searchDataForPhrase(phrase: String): LiveData<List<PostWithUser>> {
    return dbRepository.searchData(phrase)
  }

  fun getAllData(): LiveData<List<PostWithUser>> {
    return dbRepository.fetchPostsWithUser()
  }

  fun makeAlbumLiveData(userId: Long): LiveData<List<AlbumListItem>> {
    return LiveDataReactiveStreams.fromPublisher(
        dbRepository.fetchAlbumsForUser(userId)
            .filter {
              it.isNotEmpty()
            }.map {
              Mapper.unfoldList(it, AlbumWithPhotosToAlbumListItemMapper())
            }
    )
  }
}